pkg/sources/git/git.go

```diff
@@ -466,78 +483,110 @@ func (s *Git) ScanCommits(ctx context.Context, repo *git.Repository, path string
 	// get the URL metadata for reporting (may be empty)
 	urlMetadata := getSafeRemoteURL(repo, "origin")
 
-	var depth int64
+	ctx = context.WithValues(ctx, "repo", urlMetadata)
 
 	gitDir := filepath.Join(path, gitDirName)
+	meta := scanMetadata{gitDir: gitDir, urlMetadata: urlMetadata, reporter: reporter, scanOptions: scanOptions}
+	s.startWorkers(ctx, meta, commitsWorker, commitChan)
+
+	return nil
+}
+
+func (s *Git) startWorkers(ctx context.Context, meta scanMetadata, worker workerFunc, commitChan <-chan gitparse.Commit) {
+	var wg sync.WaitGroup
 
-	logger := ctx.Logger().WithValues("repo", urlMetadata)
-	logger.V(1).Info("scanning repo", "base", scanOptions.BaseHash, "head", scanOptions.HeadHash)
-	for commit := range commitChan {
-		if len(scanOptions.BaseHash) > 0 {
-			if commit.Hash == scanOptions.BaseHash {
-				logger.V(1).Info("reached base commit", "commit", commit.Hash)
-				break
+	for i := 0; i < s.concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ctx := context.WithValue(ctx, "scan_commit_worker_id", common.RandomID(5))
+			if err := worker(ctx, s, meta, commitChan); err != nil {
+				ctx.Logger().Error(err, "error in worker")
 			}
-		}
-		if scanOptions.MaxDepth > 0 && depth >= scanOptions.MaxDepth {
-			logger.V(1).Info("reached max depth", "depth", depth)
-			break
-		}
-		depth++
-		atomic.AddUint64(&s.metrics.commitsScanned, 1)
-		logger.V(5).Info("scanning commit", "commit", commit.Hash)
-		for _, diff := range commit.Diffs {
-			if !scanOptions.Filter.Pass(diff.PathB) {
-				continue
+		}()
+	}
+
+	wg.Wait()
+}
+
+func commitsWorker(ctx context.Context, s *Git, meta scanMetadata, commitChan <-chan gitparse.Commit) error {
+	for {
+		select {
+		case commit, ok := <-commitChan:
+			if !ok {
+				return nil
 			}
 
-			fileName := diff.PathB
-			if fileName == "" {
-				continue
+			var depth int64
+
+			if len(meta.scanOptions.BaseHash) > 0 {
+				if commit.Hash == meta.scanOptions.BaseHash {
+					ctx.Logger().V(1).Info("reached base commit", "commit", commit.Hash)
+					return nil
+				}
+			}
```

Comment:
Ahh yep, that's true. If we hit this we should actually signal to all the workers we are done.... wrap it up, call it a day. I'll adjust.

Reply:
Wouldn't doing it in the workers introduces a race condition, and we might scan more commits than asked?

Reply:
Yea, I think this might not be the best approach. I think we should have the workers simply handle the binary diffs. Leave the rest to be handled serially.